README
See README-alc1100


ChangeLog

EPSON AVASYS Corporation Linux Driver Team <pipsplus-bugs@avasys.jp>

	* End of file


alc1100_lprwrapper.sh
#!/bin/bash
#  alc1100_lprwrapper.sh -- LPR(ng) wrapper filter for ALC1100-ESC/PageS-filter
#  Copyright (C) SEIKO EPSON CORPORATION 2003-2006.
#
#  This file is part of the Epson-ALC1100-filter.
#
#  The Epson-ALC1100-filter is free software.
#  You can redistribute it and/or modify it under the terms of the EPSON
#  AVASYS Public License (versions dated 2005-04-01 or later) as published
#  by EPSON AVASYS Corporation.
#
#  This program is distributed in the hope that it will be useful, but
#  WITHOUT ANY WARRANTY;  without even the implied warranty of FITNESS
#  FOR A PARTICULAR PURPOSE or MERCHANTABILITY.
#  See the EPSON AVASYS Public License for more details.
#
#  You should have received a verbatim copy of the EPSON AVASYS Public
#  License along with this program; if not, write to:
#
#      EPSON AVASYS Corporation
#      1077-5 Ostu Shimonogo
#      Ueda Nagano 386-1214  JAPAN
#
# !Caution: This script work on "bash" only.
declare -a PaperList[0]
declare -a PaperIndex[0]
declare -a SizeList[0]

DEBUG="off"
#DEBUG="on"
#THROUGH="on"
DATADIR=/etc/epkowa/alc1100
TOP=/tmp
#TOP=${TMPDIR:=/tmp}

ARGS="$@"

OPTKEY=" \
    PageSize\
    InputSlot\
    Resolution\
    Duplex\
    Color\
    TonerSave\
    MediaType\
    Copies\
"

# Paper size parameters for EPSON AL-C1100
PaperList=(\
"PageSize=a4     XY600=4960x7016" \
"PageSize=b5     XY600=4300x6072" \
"PageSize=a5     XY600=3496x4960" \
"PageSize=lt     XY600=5100x6600" \
"PageSize=hlt    XY600=3300x5100" \
"PageSize=exe    XY600=4350x6300" \
"PageSize=glt    XY600=4800x6300" \
"PageSize=mon    XY600=2324x4500 Rotate=90" \
"PageSize=ib5    XY600=4156x5905" \
"PageSize=env10  XY600=2474x5700" \
"PageSize=envdl  XY600=2598x5196 Rotate=90" \
"PageSize=envc5  XY600=3826x5408" \
"PageSize=envc6  XY600=2692x3826 Rotate=90" \
)

SizeList=(\
"-w595.2 -h841.92 -W595.2 -H841.92" \
"-w516 -h728.64 -W516 -H728.64" \
"-w419.52 -h595.2 -W419.52 -H595.2" \
"-w612 -h792 -W612 -H792" \
"-w396 -h612 -W396 -H612" \
"-w522 -h756 -W522 -H756" \
"-w576 -h756 -W576 -H756" \
"-w278.88 -h540 -W278.88 -H540" \
"-w498.72 -h708.6 -W498.72 -H708.6" \
"-w296.88 -h684 -W296.88 -H684" \
"-w311.76 -h623.52 -W311.76 -H623.52" \
"-w459.12 -h648.96 -W459.12 -H648.96" \
"-w323.04 -h459.12 -W323.04 -H459.12" \
)

PaperIndex=( ${PaperList[*]/ */} )


default_opt=""
test -f $DATADIR/option.conf && default_opt="$(cat $DATADIR/option.conf | sed '/^#/d')"

# DEBUG message out
LOG=$TOP/alc1100_lprwrapper.log

test "$DEBUG" = "on" && echo $0 $ARGS > $LOG

# log message out
msglog()
{
    test "$DEBUG" = "on" && echo "$@" >> $LOG
    return 0
}

# get option value
optvalue()
{
    test -z "$1" && return 1
    KEY="$1"
    test -z "$(echo $ARGS | grep $KEY= )" && return 1
    VALUE=$(echo $ARGS | sed -e "s,^.*${KEY}=\\([^ ,\$]*\\).*,\\1,")
#    msglog "optval: $KEY=$VALUE"
    return 0
}

# get option value
default_value()
{
    test -z "$1" && return 1
    KEY="$1"
    test -z "$(echo $default_opt | grep $KEY= )" && return 1
    VALUE=$(echo $default_opt | sed -e "s,^.*${KEY}=\\([^ \$]*\\).*,\\1,")
    msglog "default: $KEY=$VALUE"
    return 0
}

#
# get paper size strings
# ex) get_paper_size "PageSize=a4"
#     return: A4 paper size setting parameter strings.
#
get_paper_size()
{
    test -z "$1" && return 1
    n=0
    while test -n "${PaperIndex[$((n))]}" \
	-a "${PaperIndex[$n]}" != "$1"
      do n=$((n +1))
    done
    paper_size="${PaperList[$n]}"
    adjust_args="${SizeList[$n]}"
    msglog "paper_size=$paper_size"
    return 0
}


# ----------------------------------------
#
# MAIN START
#

filter_args=""

for key in $OPTKEY; do
    v="" && optvalue $key && v=$VALUE && msglog "$key=$v"
    if test "$key" = "PageSize";then
	VALUE="$v" && test -z "$v" && default_value $key
	paper_size="" && adjust_args="" && get_paper_size "$key=$VALUE"
	filter_args="$filter_args $paper_size"
    else
	if test -z "$v"; then
	    default_value $key && filter_args="$filter_args $key=$VALUE"
	else
	    filter_args="$filter_args $key=$v"
	fi
    fi
done


tee0="" && test "$DEBUG" = "on" && tee0="tee $TOP/alc1100_lprwrapper.ps | "
sizeadjust="psresize -q $adjust_args | " && test x"$THROUGH" = x"on" && sizeadjust=""
filtercmd="pstoalc1100.sh $filter_args"
cmd="$tee0 $sizeadjust $filtercmd"

sh -c "$cmd"

msglog "$cmd"

#end of file


pstoalc1100.sh
#!/bin/bash
#  pstoalc1100.sh -- PostScript to ALC1100-ESC/PageS-filter
#  Copyright (C) SEIKO EPSON CORPORATION 2003-2006.
#
#  This file is part of the Epson-ALC1100-filter.
#
#  The Epson-ALC1100-filter is free software.
#  You can redistribute it and/or modify it under the terms of the EPSON
#  AVASYS Public License (versions dated 2005-04-01 or later) as published
#  by EPSON AVASYS Corporation.
#
#  This program is distributed in the hope that it will be useful, but
#  WITHOUT ANY WARRANTY;  without even the implied warranty of FITNESS
#  FOR A PARTICULAR PURPOSE or MERCHANTABILITY.
#  See the EPSON AVASYS Public License for more details.
#
#  You should have received a verbatim copy of the EPSON AVASYS Public
#  License along with this program; if not, write to:
#
#      EPSON AVASYS Corporation
#      1077-5 Ostu Shimonogo
#      Ueda Nagano 386-1214  JAPAN
#
DEBUG="off"
PSOUT="off"
WORKF="off"
TOP=/tmp
#TOP=${TMPDIR:=/tmp}
TMPFILE="$TOP/pstoalc1100"

ARGS="$@"

# for debug
#DEBUG="on"
#PSOUT="on"
#WORKF="on"

WORKFLAG="${TMPFILE}_at_work_now"
clear_work_flag=""
if test "$WORKF" = "on" ; then
    clear_work_flag=" ; test -f $WORKFLAG && rm -f $WORKFLAG"
    while test -f $WORKFLAG ; do sleep 2 ; done
    sleep 2 && touch $WORKFLAG
fi

LOG=${TMPFILE}.log
test "$DEBUG" = "on" && echo $0 $ARGS > $LOG

tee0="" && test "$DEBUG" = "on" && tee0="| tee ${TMPFILE}In.ps"
tee1="" && test "$DEBUG" = "on" && tee1="| tee ${TMPFILE}1.ps"
tee2="" && test "$DEBUG" = "on" && tee2="| tee ${TMPFILE}2.ps"
tee3="" && test "$DEBUG" = "on" && tee3="| tee ${TMPFILE}3.ps"
tee8="" && test "$DEBUG" = "on" && tee8="| tee ${TMPFILE}Out.ps"
tee9="" && test "$DEBUG" = "on" && tee9="| tee ${TMPFILE}.ras"

# get option value
optvalue()
{
    test -z "$1" && return 1
    KEY="$1"
    test -z "$(echo $ARGS | grep $KEY= )" && return 1
    VALUE=$(echo $ARGS | sed -e "s,^.*${KEY}=\\([^ ,\$]*\\).*,\\1,")
    return 0
}

# error message out
msgerr()
{
    cont=0 && test x"$1" = x"-n" && cont=1 && shift
    a="$@"
    test 0 -ne $# && echo -n "$@" >&2
    test 0 -eq $cont && echo "" >&2
    return 0
}

# log message out
msglog()
{
    test "$DEBUG" = "on" && echo "$@" >> $LOG
    return 0
}

#
# START MAIN
#

# Paper size in 600dpi
XY600="" && optvalue XY600 && XY600=$VALUE && msglog "XY600=$XY600"
test -z "$XY600" && msgerr "ERROR: $0: Un-known paper size" && exit 1

# Output resolution.
Resolution="" && optvalue Resolution && Resolution=$VALUE && msglog "Resolution=$Resolution"
test -z "$Resolution" && msgerr "ERROR: $0: Un-known output resolution" && exit 1
Resolution=$(echo $Resolution | sed -e 's,^*[^[:digit:]],,;s,[^[:digit:]]*$,,')

# other options
PageSize=""  && optvalue PageSize  && PageSize=$VALUE  && msglog "PageSize=$PageSize"
Rotate=""    && optvalue Rotate    && Rotate=$VALUE    && msglog "Rotate=$Rotate"
Duplex=""    && optvalue Duplex    && Duplex=$VALUE    && msglog "Duplex=$Duplex"
Color=""     && optvalue Color     && Color=$VALUE     && msglog "Color=$Color"
MediaType="" && optvalue MediaType && MediaType=$VALUE && msglog "MediaType=$MediaType"
TonerSave="" && optvalue TonerSave && TonerSave=$VALUE && msglog "TonerSave=$TonerSave"
InputSlot="" && optvalue InputSlot && InputSlot=$VALUE && msglog "InputSlot=$InputSlot"
Copies=""    && optvalue Copies    && Copies=$VALUE    && msglog "Copies=$Copies"
Collate=""   && optvalue Collate   && Collate=$VALUE   && msglog "Collate=$Collate"
collatefile="-"
numcopies=1

# Output Printable area width/height in 600dpi.
LogX600="$(echo $XY600 | sed 's,x.*$,,')"
LogY600="$(echo $XY600 | sed 's,^.*x,,')"
msglog "LogX600=$LogX600" "LogY600=$LogY600"

# Logical Printable area width/height in PS unit (1/72 inch).
LogX72=$(echo "scale=6; a=$LogX600*72/600; scale=0; b=a/1; scale=6; c=a-b; if(c>0.444444) b+1 else b" | bc)
LogY72=$(echo "scale=6; a=$LogY600*72/600; scale=0; b=a/1; scale=6; c=a-b; if(c>0.444444) b+1 else b" | bc)
msglog "LogX72=$LogX72" "LogY72=$LogY72"

# if hardware rotate paper then swap X and Y.
if test x"$Rotate" = x"90";then
    X600=$LogY600
    Y600=$LogX600
    X72=$LogY72
    Y72=$LogX72
else
    X600=$LogX600
    Y600=$LogY600
    X72=$LogX72
    Y72=$LogY72
fi
msglog "X600=$X600" "Y600=$Y600"
msglog "X72=$X72" "Y72=$Y72"

# output printable area in output resolution unit(dot)
Xdot=$(echo "a=$X600*$Resolution/600; scale=0; b=a/1; scale=4; c=a-b; if(c>0.4444) b+1 else b" | bc)
Ydot=$(echo "a=$Y600*$Resolution/600; scale=0; b=a/1; scale=4; c=a-b; if(c>0.4444) b+1 else b" | bc)
msglog "Xdot=$Xdot" "Ydot=$Ydot"

# Duplex and rotation
useRotator="on"
if test x"$Rotate" = x"90";then
    # default duplex = long edge
    pstops1="-w$LogY72 -h$LogY72"
    if test x"$Duplex" = x"DuplexNoTumble"; then
	pstops2="2:R1\(0,$Y72\),R0\(0,$Y72\)"
    else
	if test x"$Duplex" = x"DuplexTumble"; then
	    pstops2="2:L1\($X72,0\),R0\(0,$Y72\)"
	else
	    pstops2="1:R0\(0,$Y72\)"
	fi
    fi
else
    # default duplex = short edge
    pstops1="-w$X72 -h$Y72"
    if test x"$Duplex" = x"DuplexNoTumble"; then
	pstops2="2:1U\($X72,$Y72\),0"
    else
	if test x"$Duplex" = x"DuplexTumble"; then
	    pstops2="2:1,0"
	else
	    pstops2="1:0"
	    useRotator="off"
	fi
    fi
fi

# color/grayscale out control.
if test x"$Color" = x"black"; then
    gs1="-sDEVICE=bit -dGrayValues=256"
else
    gs1="-sDEVICE=bitrgb -dRedValues=256"
fi

gs2="-r$Resolution -g${Xdot}x${Ydot}"


# create ESC/PageS filter option argument.
opt="-o Width:${Xdot} -o Height:${Ydot}"
if test x"$Resolution" = x"300"; then
    opt="$opt -o QualityAuto:fast"
else
    opt="$opt -o QualityAuto:fine"
fi
if test x"$Duplex" = x"DuplexTumble"; then
    opt="$opt -o Duplex:top"
else
    if test x"$Duplex" = x"DuplexNoTumble"; then
    opt="$opt -o Duplex:left"
    fi
fi
test -n "$Color" && opt="$opt -o Color:$Color"
test -n "$PageSize" && opt="$opt -o PaperSize:$PageSize"
test -n "$MediaType" && opt="$opt -o PaperType:$MediaType"
test -n "$InputSlot" && opt="$opt -o PaperSource:$InputSlot"
test -n "$TonerSave" && opt="$opt -o TonerSave:$TonerSave"
if test 1 -ge $((Copies)); then
    Copies="0"
else
    if test "$Collate" = "on"; then
	collatefile="$( mktemp $TOP/alc1100.XXXXXX )"
	cat > $collatefile
	numcopies=$((Copies))
    else
	test -n "$Copies" && opt="$opt -o Copies:$Copies"
    fi
fi

msglog "collatefile=$collatefile"
msglog "numcopies=$numcopies"

rotator="" && test "$useRotator" = "on" && rotator="| pstops -q $pstops1 $pstops2"
rasterizer="| gs -q -dBATCH -dPARANOIDSAFER -dNOPAUSE $gs1 $gs2 -sOutputFile=- -"
escpagesfilter="| alc1100 $opt" && test "$PSOUT" = "on" && escpagesfilter=""

cmd="\
cat $collatefile \
$tee0 $rotator \
$tee8 $rasterizer \
$tee9 $escpagesfilter \
$clear_work_flag"


# execute transration.
page=1
while test 0 -ne $((numcopies));do
    msglog "page=$page numcopies=$numcopies"
    sh -c "$cmd"
    numcopies=$((numcopies -1))
    page=$((page +1))
done

test $collatefile != "-" && rm -f $collatefile

# for debug
msglog "[ Command ]"
msglog "$cmd"
msglog "[ Emulation ]"
msglog "cat ${TMPFILE}In.ps $rotator > d1.ps"
msglog "cat d1.ps $rasterizer > d2.ras"
msglog "cat d2.ras $escpagesfilter > d3.prn"
msglog "[ View ]"
msglog "gs -q -r72 -g${LogX72}x${LogY72} -dBATCH -sDEVICE=x11 ${TMPFILE}In.ps"
msglog "gs -q -r72 -g${X72}x${Y72} -dBATCH -sDEVICE=x11 ${TMPFILE}Out.ps"
if test x"$Color" = x"black"; then
    TYPE="Gray"
else
    TYPE="RGB"
fi
msglog "convert -depth 8 -size ${Xdot}x${Ydot} $TYPE:${TMPFILE}.ras pstoalc1100.png"
msglog "done"

exit 0
# end of script
